File: python/queue.py
# Queue with linked list
class Node:
    def __init__(self, item):
        self.item = item
        self.next = None


class Queue:
    def __init__(self):
        self.head = None
        self.tail = None
        self._length = 0

    def enqueue(self, item):
        new_node = Node(item)

        if self.is_empty():
            self.head = self.tail = new_node

        else:
            self.tail.next = new_node
            self.tail = new_node

        self._length += 1

    def deque(self):
        if self.is_empty():
            return None

        temp = self.head

        if self._length == 1:
            self.head = self.tail = None
        else:
            self.head = self.head.next

        self._length -= 1

        return temp.item

    def is_empty(self):
        return self.head is None

    def peek(self):
        return self.head.item or None

    def length(self):
        return self._length


# ===========================
# Queue using Python's List

class Queue_PyList:
    def __init__(self):
        self.queue = []

    def enqueue(self, item):
        self.queue.append(item)

    def deque(self, item):
        if self.is_empty():
            return None

        return self.queue.pop(0)

    def peek(self):
        return self.queue[0]

    def is_empty(self):
        return len(self.queue) == 0

    def length(self):
        return len(self.queue)

File: typescript/queue.ts
type QNode<T> = {
    item: T,
    next?: QNode<T>,
}

class Queue<T> {
    public length: number;
    private head?: QNode<T>;
    private tail?: QNode<T>;

    constructor() {
        this.length = 0;
        this.head = this.tail = undefined;
    }

    enqueue(item: T): void {
        const newNode: QNode<T> = { item: item }

        this.length++;
        if (!this.head) {
            this.head = this.tail = newNode;
        } else {
            this.tail.next = newNode;
            this.tail = newNode;
        }
    }

    deque(): T | undefined {
        if (!this.head) {
            return undefined;
        }

        this.length--;
        const item = this.head?.item
        this.head = this.head.next
        return item;
    }

    peek(): T | undefined {
        return this.head?.item;
    }
}

function test() {
    const newQueue = new Queue();
    console.log('length', newQueue.length);

    newQueue.enqueue('foo');
    console.log('\nenqueued foo');
    console.log('length', newQueue.length);
    console.log('peek', newQueue.peek());

    newQueue.enqueue('bar');
    console.log('\nenqueued bar');
    console.log('length', newQueue.length);
    console.log('peek', newQueue.peek());

    newQueue.enqueue(1);
    console.log('\nenqueued 1');
    console.log('length', newQueue.length);
    console.log('peek', newQueue.peek());

    newQueue.deque();
    console.log('\ndequed');
    console.log('length', newQueue.length);
    console.log('peek', newQueue.peek());

    newQueue.deque();
    console.log('\ndequed');
    console.log('length', newQueue.length);
    console.log('peek', newQueue.peek());

    newQueue.deque();
    console.log('\ndequed');
    console.log('length', newQueue.length);
    console.log('peek', newQueue.peek());
}
test();
